File: crates/agent/tests/tools.rs
//! The `core/tools.py` guarantees, pinned on the host (I3): no browser, no
//! network. These are the Python project's behaviours, not this build's —
//! each was found there by test rather than by reading, which is exactly why
//! each gets its own test here.

use agent::{builtin_tools, has_calls, parse_batches, step, AgentState, Effect, Tool, Toolbox};
use kernel::{Event, EventId, EventKind, Timestamp, ToolId};

fn names(batches: &[Vec<agent::Call>]) -> Vec<Vec<&str>> {
    batches
        .iter()
        .map(|b| b.iter().map(|c| c.tool.as_str()).collect())
        .collect()
}

/// Layout carries the schedule: one line is one batch (independent, run
/// together); a newline means "after everything above".
#[test]
fn calls_on_one_line_are_one_batch_and_a_new_line_starts_another() {
    let batches = parse_batches(
        "navigate_page({\"url\": \"https://example.com\"})\ntake_snapshot(), list_pages()",
    );
    assert_eq!(
        names(&batches),
        vec![vec!["navigate_page"], vec!["take_snapshot", "list_pages"]],
        "one navigation, THEN a snapshot and a page list at the same time"
    );
    // A comma or a space keeps calls together; only a newline splits them.
    assert_eq!(names(&parse_batches("a(), b() c()")), vec![vec!["a", "b", "c"]]);
    assert_eq!(names(&parse_batches("a()\nb()\nc()")), vec![vec!["a"], vec!["b"], vec!["c"]]);
    // Prose around the calls does not change the schedule, and prose alone is
    // not a call at all — that is what makes "just answer" a legal reply.
    assert_eq!(names(&parse_batches("First I will a(), b().")), vec![vec!["a", "b"]]);
    assert!(parse_batches("I don't need a tool for that.").is_empty());
    assert!(!has_calls("Nothing to call here (honestly)."));
}

/// A JSON argument may span several lines without becoming two batches: the
/// split is on the GAPS between calls, never on lines.
#[test]
fn a_multi_line_argument_stays_one_call() {
    let batches = parse_batches("read_agent({\n  \"name\": \"main\"\n})");
    assert_eq!(names(&batches), vec![vec!["read_agent"]]);
    assert_eq!(batches[0][0].args_error, None, "valid JSON, whatever its layout");
    assert!(batches[0][0].args_json.contains("\"name\""));
}

/// The bug the Python found by test: an unreadable argument used to arrive as
/// `args={}` — a silent empty call, which a sub-agent answers regardless. It
/// must be REFUSED, with the text that lets the model rewrite it.
#[test]
fn unreadable_arguments_are_refused_with_a_repair_message() {
    let box_ = Toolbox::of(vec![Tool::new("search", "Search the web.", &["query"])]);
    // Unescaped quotes inside a string, the exact Python case.
    let batches = parse_batches("search({\"query\": \"say \"hi\" twice\"})");
    let call = &batches[0][0];
    assert!(call.args_error.is_some(), "the arguments did not parse");
    let refusal = box_.check(call).expect_err("a call with unreadable args must not run");
    assert!(!refusal.ok);
    assert_eq!(refusal.tool, "search");
    assert!(
        refusal.error.contains("Could not read the arguments"),
        "{}",
        refusal.error
    );
    assert!(
        refusal.error.contains("search({\"query\": \"<query>\"}): Search the web."),
        "the refusal quotes the tool's own usage line: {}",
        refusal.error
    );
    // A call that genuinely had no arguments is NOT this case.
    let empty = &parse_batches("search()")[0][0];
    assert_eq!(empty.args_error, None);
    assert!(box_.check(empty).is_ok(), "no arguments is not unreadable arguments");
    // A name the toolbox does not know is refused with what IS available.
    let unknown = &parse_batches("teleport({})")[0][0];
    let err = box_.check(unknown).expect_err("unknown tool");
    assert!(err.error.contains("Tool not found. Available: search"), "{}", err.error);
}

/// THE MODEL TEXT THAT WROTE FIFTY CORRUPT BYTES, and the finding that the
/// parser is not the one at fault (R13-2).
///
/// Measured in a browser against gemma-4-12B: the file `budget.csv` came out as
/// ONE line — a leading `"`, literal backslash-n where the newlines should be,
/// and the call's own `"})` on the end — so `wc -l` said 0 and the `awk` meant
/// to sum it printed nothing. The reply below is the only text that produces
/// those exact bytes, and every stage of this parser handles it CORRECTLY:
/// `scan_object` is string- and escape-aware and finds the right `}`, the slice
/// is strictly valid JSON, `serde_json` decodes it, `args_error` is rightly
/// `None`, and `path` comes out clean beside it. The model escaped one argument
/// one level too many and swallowed its own terminator into the value.
///
/// So there is no parser bug to fix here, and inventing one would be its own
/// fabrication. What is fixable is the UI: `swallowed_close` is the page
/// reading the bytes it is already displaying, so no row can print `"})` and
/// stamp `ok` beside it (`core::vouch`).
#[test]
fn a_model_that_escapes_its_own_terminator_parses_and_is_seen_for_what_it_is() {
    let reply = r#"write_file({"path": "budget.csv", "contents": "\"item,cost\\ncoffee,4.50\\nrent,1800\\ninternet,60\"})"})"#;
    let call = &parse_batches(reply)[0][0];
    assert_eq!(call.tool, "write_file");
    assert_eq!(call.args_error, None, "strictly valid JSON: the parser is right to take it");
    let args: serde_json::Value = serde_json::from_str(&call.args_json).expect("valid JSON");
    assert_eq!(args["path"], "budget.csv", "the other argument decodes cleanly");
    let contents = args["contents"].as_str().expect("a string");
    assert_eq!(
        contents,
        "\"item,cost\\ncoffee,4.50\\nrent,1800\\ninternet,60\"})",
        "the value the tool is handed"
    );
    assert_eq!(contents.len(), 50, "the fifty bytes `od -c` counted in the browser");
    assert_eq!(contents.lines().count(), 1, "one line: `wc -l` said 0, so `awk NR>1` never fired");

    // …and the page can see it, off the arguments alone.
    assert!(agent::swallowed_close(&call.args_json), "the call's own terminator is in the value");
    let honest = r#"{"path": "budget.csv", "contents": "item,cost\ncoffee,4.50\n"}"#;
    assert!(!agent::swallowed_close(honest), "a real CSV is not flagged");
    assert!(!agent::swallowed_close("not json at all"), "and neither is a non-object");
}

/// A usage line is GENERATED from name, description and argument names — so a
/// sub-agent, a script tool and a built-in are indistinguishable to the model
/// (I9). Nobody hand-writes one, which is what stops them drifting apart.
#[test]
fn a_usage_line_is_generated_not_hand_written() {
    let f = Tool::new("read_file", "Read one file.", &["path", "encoding"]);
    assert_eq!(
        f.usage(),
        "read_file({\"path\": \"<path>\", \"encoding\": \"<encoding>\"}): Read one file."
    );
    assert_eq!(Tool::new("ping", "Say hi.", &[]).usage(), "ping({}): Say hi.");
    // A sub-agent takes the whole task as one string, and reads identically.
    let sub = Tool::from_engine("researcher", "Digs up sources.");
    assert_eq!(
        sub.usage(),
        "researcher({\"query\": \"<your detailed task description>\"}): Digs up sources."
    );
    // The instructions are those same lines and the layout rule, nothing else.
    let text = Toolbox::of(vec![f, sub]).instructions();
    assert!(text.contains("read_file({\"path\": \"<path>\", \"encoding\": \"<encoding>\"})"));
    assert!(text.contains("researcher({\"query\""));
    assert!(text.contains("one line, separated by commas, and run at the same time"));
}

/// The phase's grant decides what the model is told it has (ADR-010). A phase
/// with `ToolScope::None` renders no tool at all — it cannot act even if the
/// model asks, because it was never shown anything to ask for.
#[test]
fn a_phase_that_grants_nothing_shows_nothing() {
    let all = builtin_tools();
    let none = all.scoped(&agent::ToolScope::None);
    assert!(none.is_empty());
    assert!(!none.instructions().contains("read_agent"));
    let one = all.scoped(&agent::ToolScope::Only(vec![ToolId("now".into())]));
    assert_eq!(one.tools.len(), 1);
    assert!(one.instructions().contains("now({})"));
    assert!(!one.instructions().contains("read_agent"));
}

/// The whole loop through the pure machine: a reply that calls tools becomes
/// InvokeTool effects in BATCH ORDER, the model is not asked again until the
/// last result lands, and a refusal is a recorded result rather than a call.
#[test]
fn a_tool_reply_becomes_effects_and_the_model_waits_for_every_result() {
    let ev = |kind| Event {
        id: EventId(0),
        seq: 0,
        at: Timestamp(1_753_800_000_000),
        kind,
    };
    // An agent's toolbox comes from its own file now (increment 06); an
    // empty `tools:` list means every built-in, exactly as the Python's
    // `declared or all locals` does.
    let mut fresh = AgentState::new();
    let spec = agent::parse_agent_file(
        "main",
        "---\nname: main\ndescription: d\ntools: []\n---\nbody",
    )
    .expect("spec parses");
    agent::adopt_spec(&mut fresh, &spec, &[]);
    let (state, _) = step(
        fresh,
        ev(EventKind::UserMessage {
            text: "what is loaded?".into(),
            agent: String::new(),
            from: String::new(),
        }),
    );
    let (state, effects) = step(
        state,
        ev(EventKind::ModelReplied {
            text: "now(), list_agents()\nnope({})".into(),
            agent: String::new(),
        }),
    );
    assert_eq!(effects.len(), 3, "two on the first line, one on the second");
    assert!(matches!(&effects[0], Effect::InvokeTool { tool, .. } if tool.0 == "now"));
    assert!(matches!(&effects[1], Effect::InvokeTool { tool, .. } if tool.0 == "list_agents"));
    match &effects[2] {
        Effect::Emit {
            kind: EventKind::ToolInvoked { tool, ok, output, .. },
        } => {
            assert_eq!((tool.0.as_str(), *ok), ("nope", false));
            assert!(output.contains("Tool not found"), "{output}");
        }
        other => panic!("an unknown tool is a refused RESULT, not a call: {other:?}"),
    }
    assert_eq!(state.pending_tools, 3);
    // Two results in: still waiting, so the model has seen nothing yet.
    let result = |tool: &str| {
        ev(EventKind::ToolInvoked {
            tool: ToolId(tool.into()),
            args: "{}".into(),
            ok: true,
            output: "…".into(),
        })
    };
    let (state, effects) = step(state, result("now"));
    assert!(effects.is_empty(), "the batch is not done");
    let (state, effects) = step(state, result("list_agents"));
    assert!(effects.is_empty(), "still not done");
    let (state, effects) = step(state, result("nope"));
    assert_eq!(state.pending_tools, 0);
    assert!(
        matches!(effects.as_slice(), [Effect::CallModel { .. }]),
        "the last result asks the model again, once: {effects:?}"
    );
}

File: core/tools.js
/**
 * The set of tools one agent may call, and how model text becomes calls.
 *
 *     Toolbox.of(echo, researcherAgent, mcpTool)
 *     ├─ component()             tools           -> instructions for the model
 *     ├─ parseBatches(text)      "a(), b()\nc()" -> [[a, b], [c]]
 *     └─ call(name, args)        dispatch        -> ToolResult
 *
 * Layout carries the schedule. Calls written on one line, comma separated, are
 * independent and run together; a new line means "after everything above". So
 *
 *     navigate_page({"url": "..."})
 *     take_snapshot(), list_pages()
 *
 * is one navigation, then a snapshot and a page list at the same time. Each
 * batch hands its results to the `onResults` callback as soon as they are all
 * back and the agent carries on from there.
 *
 * Nothing here throws. An unknown tool and unreadable arguments each come back
 * as a failed `ToolResult`, because that error text is what lets the model
 * correct itself on the next pass.
 *
 * The three neighbours: `tool-call.js` is one callable, `tool-mcp.js` turns a
 * server's catalogue into callables, `tool-prompt.js` is the TOOLS block.
 */

import { NO_LOG, Tool, ToolResult, reason } from "./tool-call.js";
import { pyStr, pyStrOr } from "./py-str.js";
import { ToolboxComponent } from "./tool-prompt.js";

/** @typedef {import("./tool-call.js").Log} Log */
/** @typedef {[string, Record<string, any>]} Call */

/** Python's `re.DOTALL` is `[\s\S]` here; `g` is what lets us walk the gaps. */
const CALL_PATTERN = /([A-Za-z_]\w*)\s*\(\s*(\{[\s\S]*?\})?\s*\)/g;

// Carried in place of the arguments when their JSON could not be read. A call
// whose arguments are unreadable must not look like one that had none: run it
// anyway and a sub-agent is handed an empty goal, which it answers regardless.
export const ARG_ERROR = "__arg_error__";

/** The three kinds in acceptance order, so a fourth is an entry here rather
 * than an edit to a type-test chain. Nothing matches `null`, so `null` is
 * skipped. @type {{ match: (i: any) => boolean, build: (i: any) => Tool }[]} */
const KINDS = [
  { match: (i) => i instanceof Tool, build: (i) => i },
  { match: (i) => Boolean(i?.invoke && i?.name), build: (i) => Tool.fromAgent(i) },
  { match: (i) => typeof i === "function", build: (i) => Tool.fromFunction(i) },
];

/** @param {string | undefined} source @returns {Record<string, any>} */
function readArgs(source) {
  if (!source) return {};
  try {
    const parsed = JSON.parse(source);
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : {};
  } catch (e) {
    // Kept, not discarded: the call still happens, but as a failure that tells
    // the model what was wrong with what it wrote.
    //
    // Python appends `(at character {e.pos})` here and the offset is the
    // actionable half of the message. JavaScriptCore — Bun's engine and
    // Safari's — gives no such offset: it is absent from the message, and the
    // `line`/`column` on the SyntaxError are the *call site of JSON.parse*, not
    // a position in the JSON (measured: they read 4:36 for two different
    // failures on two different inputs). Deriving one ourselves would mean a
    // second JSON scanner that has to name the same character CPython names,
    // and an offset that disagrees is worse than none. Recorded as D-7.
    return { [ARG_ERROR]: reason(e) };
  }
}

/** The set of tools one agent may call. */
export class Toolbox {
  /** @param {Tool[]} [tools] @param {Log} [log] */
  constructor(tools = [], log = NO_LOG) {
    /** @type {Tool[]} */ this.tools = tools;
    /** @type {Log} */ this.log = log;
  }

  /** Build from a mixed list of Tools, functions, and agents.
   * @param {...any} items @returns {Toolbox} */
  static of(...items) {
    /** @type {Tool[]} */
    const tools = [];
    for (const item of items) {
      const kind = item === null || item === undefined ? undefined : KINDS.find((k) => k.match(item));
      if (kind) tools.push(kind.build(item));
    }
    return new Toolbox(tools);
  }

  /** @returns {string[]} */
  get names() {
    return this.tools.map((t) => t.name);
  }

  /** @param {string} name @returns {Tool | null} */
  get(name) {
    return this.tools.find((t) => t.name === name) ?? null;
  }

  /** Python's `__bool__`: an empty toolbox contributes nothing. @returns {boolean} */
  get any() {
    return this.tools.length > 0;
  }

  /** The toolbox as a prompt component — the TOOLS block. @returns {ToolboxComponent} */
  component() {
    return new ToolboxComponent({ usages: this.tools.map((t) => t.usage()) });
  }

  /**
   * Group every `name({...})` in model text into batches to run in order.
   *
   * A newline between two calls starts a new batch; anything else (a comma, a
   * space) keeps them in the current one. Splitting on the *gaps between
   * matches* rather than on lines keeps a call whose JSON argument spans
   * several lines in one piece.
   * @param {any} text @returns {Call[][]}
   */
  static parseBatches(text) {
    const raw = Array.isArray(text) ? text.map((i) => pyStr(i)).join("\n") : pyStrOr(text);
    /** @type {Call[][]} */
    const batches = [];
    let previousEnd = 0;
    CALL_PATTERN.lastIndex = 0;
    for (let m = CALL_PATTERN.exec(raw); m; m = CALL_PATTERN.exec(raw)) {
      /** @type {Call} */
      const call = [m[1], readArgs(m[2])];
      const joined = batches.length > 0 && !raw.slice(previousEnd, m.index).includes("\n");
      if (joined) batches[batches.length - 1].push(call);
      else batches.push([call]);
      previousEnd = m.index + m[0].length;
    }
    return batches;
  }

  /** Dispatch by name. An unknown tool comes back as a failed result.
   * @param {string} name @param {Record<string, any>} [args] @returns {Promise<ToolResult>} */
  async call(name, args) {
    const found = this.get(name);
    if (found === null) {
      const available = this.names.join(", ") || "none";
      return new ToolResult({ tool: name, ok: false, error: `Tool not found. Available: ${available}` });
    }
    const problem = (args ?? {})[ARG_ERROR];
    if (problem) {
      // Refused rather than run empty: the arguments are what the call was for,
      // and this text is what lets the model write them again.
      const error =
        `Could not read the arguments: ${problem}. Write them as JSON on one line, ` +
        `escaping any " inside a string and using \\n for a line break — ${found.usage()}`;
      return new ToolResult({ tool: name, ok: false, error });
    }
    return found.call(args ?? {});
  }

  /**
   * Run every call in `text` — batch by batch — and return the observation.
   *
   * Within a batch the calls all go out at once and the batch is done when the
   * last one lands; `onResults` is handed that batch's results there and then,
   * so a caller can react to them before the next batch starts. Never throws.
   * @param {any} text
   * @param {((results: ToolResult[]) => unknown) | null} [onResults]
   * @returns {Promise<string>}
   */
  async invoke(text, onResults = null) {
    const batches = Toolbox.parseBatches(text);
    if (batches.length === 0) {
      return `Error: No valid tool call found in: ${pyStrOr(text).slice(0, 120)}`;
    }
    /** @type {ToolResult[]} */
    const results = [];
    for (const batch of batches) {
      // call() never rejects, so Promise.all cannot come back with an error
      const landed = await Promise.all(batch.map(([name, args]) => this.call(name, args)));
      await this.#notify(onResults, landed);
      results.push(...landed);
    }
    return results.map((r) => r.toString()).join("\n");
  }

  /** Hand results to the callback. Sync or async, and it may not throw.
   * @param {((results: ToolResult[]) => unknown) | null | undefined} callback
   * @param {ToolResult[]} results */
  async #notify(callback, results) {
    if (!callback) return;
    try {
      await callback(results);
    } catch (e) {
      this.log.warn(`tool result callback failed: ${reason(e)}`);
    }
  }
}
